feat: implement StudentEmail component with validation and unit tests

Charan-devv opened the pull request:
## Overview

This pull request introduces the `StudentEmail` component, which allows composing emails to students. The component supports student lookup by ID, displays student details, and provides validation for email content.

---

## Features Implemented

- **StudentEmail Component**
  - Renders student profile details (name, email, notes, image)
  - Textarea for composing email content
  - "Compose Email" button opens a pre-filled mailto link
  - Validation to ensure email content is not empty

- **Validation**
  - Displays alert if no content is entered before composing


- **Type Safety**
  - Uses the `Student` type for props and state

---

## Tests Added

- Render case when no student exists (`Student not found.`)
- Render case when student exists (profile details displayed)
- Validation test to ensure alert shows when composing email without content

---

## Related Work

- `findStudentById` implementation is available in PR `feat: implement student lookup with validation and test coverage`
- `Student` type definition is available in PR `fix: add Email value object with validation and tests`

## GitHub Issues
- https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/436
- https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/issues/437

Commit:
feat: add StudentEmail component for composing emails to students

- Implement StudentEmail React component with:
  - Lookup of student by ID using
  - Display of student profile details and notes
  - Email content input and mailto link generation with validation
- Use  type for type safety and props definition

Note:
- Implementation of  is available in PR
-  type definition is available in PR

Commit:
test: add unit test for StudentEmail component to verify missing student case

- Verify that 'Student not found.' message renders when no student data is provided

Commit:
test: add unit test for StudentEmail to verify rendering of student details

- Ensure student name, email, and notes are displayed when student data exists

Commit:
test: add validation test for StudentEmail empty content alert

- Verify alert is shown when composing email without entering content

achow-conestogac commented on src/Enrollment/Student/Pages/StudentEmail/StudentEmail.tsx:
```diff
@@ -0,0 +1,73 @@
+const emptyGetter = () => [];
+
+const StudentEmail: React.FC<{ getter?: () => Student[] }> = ({ getter = emptyGetter }) => {
+  const { id } = useParams();
+
+  const student = id ? findStudentById(id, getter) : undefined;
```
i do not understand this.  `useParams` gives you a randomized id.

why are you looking for students by a randomize value?

This is actually a symptom of writing dead code (code that is not being used anywhere).  I fail to see how this is a useful implementation.

Please include the usage of these functions as part of the PR.

Commit:
refactor(StudentEmail): remove useParams and findStudentById
- Component now requires a Student prop.
- Cleaned up mailto template literals and alt text.

Commit:
test(StudentEmail): update tests to remove routing dependency and use student prop

achow-conestogac commented:
please use this

Charan-devv commented:
[cid:5d122fb8-a7e4-454f-b516-d39c34f642e8]Hi Professor,
Sorry I can't understand this! You mentioned as "please use this". Can you please give me specific detail on what do I need to update or you saying to resolve the conversation and merge?
Thank you.
________________________________
From: Andy Chow ***@***.***>
Sent: July 14, 2025 10:05 PM
To: conestogac-acsit/SENG8130-Software-Quality-Applications-Lab ***@***.***>
Cc: Charankanth Reddy Chinthala ***@***.***>; Author ***@***.***>
Subject: Re: [conestogac-acsit/SENG8130-Software-Quality-Applications-Lab] feat: implement StudentEmail component with validation and unit tests (PR #428)

Caution: This email originated from outside of the organization. Do not click links or open attachments unless you recognize the sender and know the content is safe. Please send any questions or concerns to ***@***.***

[https://avatars.githubusercontent.com/u/168104245?s=20&v=4]achow-conestogac left a comment (conestogac-acsit/SENG8130-Software-Quality-Applications-Lab#428)<https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/pull/428#issuecomment-3071651456>

please use this

—
Reply to this email directly, view it on GitHub<https://github.com/conestogac-acsit/SENG8130-Software-Quality-Applications-Lab/pull/428#issuecomment-3071651456>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/BOJHX6BQUBCUEDL63WQ7BUT3IROVNAVCNFSM6AAAAACAYN6YOGVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZTANZRGY2TCNBVGY>.
You are receiving this because you authored the thread.Message ID: ***@***.***>

________________________________
Conestoga Official Disclaimer: This email and any attachments are for the sole use of the intended recipients and may be privileged or confidential. Any distribution, printing or other use by anyone else is prohibited. If you are not an intended recipient, please contact the sender immediately, and permanently delete this email and attachments.

Commit:
chore: ensure the branch is up to date with main

Commit:
refactor: make StudentEmail component independent by supporting optional student prop

Commit:
feat: add routing for StudentEmail page

Commit:
chore: Updated the app.tsx to remove the unused code

Commit:
Refactor: Moved the router of student email

Commit:
feat: add the appropriate index file

Commit:
Merge branch 'main' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/student_email

Commit:
Updated the current branch as per the main

Commit:
chore: revert unintended changes to package-lock.json

Commit:
feat: add nested route for StudentEmail under /enrollment

Commit:
refactor: apply dependency inversion for StudentEmail to support testing

Commit:
test: add button click test for StudentEmail with injected mailto handler

Charan-devv commented on src/App.tsx:
```diff
@@ -1,11 +1,18 @@
     <Router>
       <Routes>
+        <Route
+          path="/student-email"
```
Hello Professor,
Thankyou for you feedback.
Update the code and the now the etudent-email route is nested within the 'enrollment' route.

Charan-devv commented on src/Enrollment/Student/Pages/StudentEmail/StudentEmail.test.tsx:
Hello professor,
Thankyou for the suggestion.
I’ve updated the StudentEmail component to address the testability issue by applying dependency inversion. Specifically, I replaced the direct usage of window.location.href with an injectable onComposeEmail prop.

This allows the behavior to be fully testable, and I’ve added a new test case to verify that the Compose Email button generates the correct mailto: link when clicked.

Please let me know if you'd like any further changes to the structure or test coverage.

Commit:
Added the ourlet within the Dashboard so that navigating to the enrollemnt/student-email doesnot throw any issue

Charan-devv commented:
Hello professor,
With the current changes, the issue is resolved.
<img width="1919" height="967" alt="image" src="https://github.com/user-attachments/assets/061c4174-62f2-4f36-8951-620af8605245" />

Commit:
removed the / from the path student-email

Commit:
Removed the un-necessary comments

Commit:
added the user icon png

Commit:
Refactor: HAndeled the student profile icon as well as the button is now used from the main component

Charan-devv commented on src/Enrollment/Student/Pages/StudentEmail/StudentEmail.tsx:
```diff
@@ -0,0 +1,56 @@
+  const [content, setContent] = useState("");
+  const handleComposeEmail = useCallback(() => {
+    if (!content.trim()) {
+      alert("Please enter email content before composing.");
```
The alert is being handled.

Charan-devv commented on src/Enrollment/Student/Pages/StudentEmail/StudentEmail.tsx:
```diff
@@ -0,0 +1,63 @@
+      <div className="bg-white shadow rounded-lg p-6">
+        <div className="flex items-center gap-4 mb-6">
+          <img
+            src={student?.imageUrl}
```
Add a user icon as a placeholder so that if the user goes offline he/she will see the user icon.

Commit:
Updated the Dashbord.tsx so that only Outlet is there

Commit:
Merge branch 'main' of github.com:conestogac-acsit/SENG8130-Software-Quality-Applications-Lab into feat/student_email

Charan-devv commented on src/Enrollment/Dashboard/Dashboard.tsx:
Updated it to contain the Outlet only.

Commit:
Added the test for error when the email content is empty

Charan-devv commented on src/Enrollment/Student/Pages/StudentEmail/StudentEmail.tsx:
```diff
@@ -0,0 +1,63 @@
+import React, { useState, useCallback } from "react";
+import { Student } from "../../studentData/studentTypes";
+
+type Props = {
+  student?: Student;
+  onComposeEmail?: (url: string) => void;
+};
```
@achow-conestogac 
Thank you for your feedback! You're absolutely right that the parameters in the StudentEmail component should not be null. Currently, to make the component independent and functional for now, I made the student and onComposeEmail optional with ?. This will allow the page to render even if the data isn't passed immediately. However, I do agree that the correct flow, when everything is connected, will involve passing the student information from the student list to the StudentEmail component once the user selects a student.

For now, I kept it independent to simplify the testing and ensure functionality in isolation, but I agree that once integrated with the student list, the student and onComposeEmail should be mandatory.

achow-conestogac replied:
yeah... you have a planning problem.

this is the type of placeholder/mocking habit that we are avoiding.  what you put in here should be a functional version... not something which requires tweaking later on.

you need to plan this out a bit more and get the correct pieces in place, remove the placeholders, and then implement this into the right place in the flow.

Charan-devv replied:
I’ve made the changes by removing placeholders and ensuring the component is fully functional, properly integrated into the flow, and aligned with the correct data handling.

Commit:
Refactor StudentEmail component to require student and onComposeEmail props

- Made 'student' and 'onComposeEmail' props mandatory to ensure proper functionality.
- Added error handling for missing 'student' or 'onComposeEmail' props.
- Updated component to gracefully handle missing data and display a fallback message.

Commit:
Add tests for missing student or onComposeEmail props in StudentEmail component

- Added test case to verify error handling when 'student' or 'onComposeEmail' is missing.
- Ensured the component displays an appropriate error message for missing props.
- Updated existing tests to match the new mandatory prop requirements for 'student' and 'onComposeEmail'.

Charan-devv commented:
Hello Professor @achow-conestogac 
Can you please review the updated changes?

Thankyou
